### PR TITLE
Run tasks on the main thread

### PR DESCRIPTION
**Pull request by zanieb**
This is a new version of #9855 rebased onto [`main`](https://github.com/PrefectHQ/prefect/tree/main) now that the upstream issue in greenback (https://github.com/oremanj/greenback/issues/22) preventing this from working with ephemeral servers is resolved. This no longer requires ugly hacks like https://github.com/PrefectHQ/prefect/pull/9870 to avoid greenlet contamination. The following description is mostly copied from the old pull request. I have no idea if this would interact with or is relevant for new work regarding "autonomous" tasks e.g. https://github.com/PrefectHQ/prefect/pull/11779

-----

When feasible, task runs are now executed on the main thread instead of a worker thread. Previously, all task runs were run in a new worker thread. This should significantly improve performance in asynchronous and sequential use cases. This allows objects to be passed to and from tasks without worrying about thread safety unless you have opted into concurrency. For example, an HTTP client or database connection can be shared between a flow and its tasks now (unless synchronous concurrency is being used).

Tasks are run in the main thread when:
- The task and flow are both asynchronous
- The flow is synchronous and the task run is sequential

Moving task execution should cause _deadlocks_ when asynchronous and synchronous tasks are mixed in an asynchronous flow because if a synchronous task depends on an asynchronous task it will block the event loop waiting for the upstream to finish but the upstream cannot run since it is now on the same event loop. For example:

```python
@flow
async def foo():
    upstream = await async_task.submit(42)  # Running in the background on the main thread event loop
    sync_task(upstream) # Blocks the event loop because there's no `await`
    # If the upstream is not done when `sync_task` is called, we deadlock

await foo()
```

To resolve these deadlocks, `greenback` is added as a dependency. `greenback` inserts a shim that lets us execute the synchronous task without blocking the event loop. This is not necessarily a big deal because we are already using SQLAlchemy which performs a similar hack with gevent to created mixed asynchronous and synchronous interfaces. However, it does add some complexity. With the shim, we no longer block the event loop when synchronous tasks are called from asynchronous flows which resolves this issue _and_ addresses general performance concerns when calling synchronous tasks from asynchronous flows (i.e. the synchronous tasks no longer block the event loop and delay asynchronous tasks). We also improve performance by spawning less threads.

In https://github.com/PrefectHQ/prefect/pull/11930/commits/a25b8b1aab3a169b5d97a6b88986947f9a437399, we also run asynchronous subflows in synchronous flows on the main thread. This was the easiest way to resolve a regression introduced by the prior changes, as well as a sensible way to deal with scheduling this case in the first place. Since an asynchronous subflow in a synchronous flow must always block until completion, it doesn't really make sense to schedule it in a worker thread.

This is a **breaking change** although I'm uncertain what kind of negative impact it would have on practical use cases. It seems likely that the usability benefits and reduction of complexity for users without concurrency should outweigh potential downsides. I think in the majority of cases it should be _more_ permissive. I'm certainly curious to hear opinions though.

Closes https://github.com/PrefectHQ/prefect/issues/9412
Closes https://github.com/PrefectHQ/prefect/issues/10627

## Example

```python
# Where we note that tasks run on the main thread, they previously ran on separate worker threads

import asyncio
import threading

from prefect import flow, task


@task
def sync_task():
    print(threading.current_thread())


@task
async def async_task():
    print(threading.current_thread())


@flow(log_prints=True)
def sync_flow():
    print(threading.current_thread())

    sync_task()  # Runs on main thread
    async_task()  # Runs on main thread
    sync_task.submit()  # Runs on worker thread
    async_task.submit()  # Runs on worker thread


@flow(log_prints=True)
async def async_flow():
    print(threading.current_thread())

    sync_task()  # Runs on worker thread
    sync_task.submit()  # Runs on worker thread
    await async_task()  # Runs on main thread
    await async_task.submit()  # Runs on main thread


sync_flow()
print("---")
asyncio.run(async_flow())
```

```
❯ python example.py
12:34:50.735 | INFO    | prefect.engine - Created flow run 'organic-wallaby' for flow 'sync-flow'
12:34:50.799 | INFO    | Flow run 'organic-wallaby' - <_MainThread(MainThread, started 8725699392)>
12:34:50.814 | INFO    | Flow run 'organic-wallaby' - Created task run 'sync_task-0' for task 'sync_task'
12:34:50.814 | INFO    | Flow run 'organic-wallaby' - Executing 'sync_task-0' immediately...
12:34:50.843 | INFO    | Task run 'sync_task-0' - <_MainThread(MainThread, started 8725699392)>
12:34:50.856 | INFO    | Task run 'sync_task-0' - Finished in state Completed()
12:34:50.870 | INFO    | Flow run 'organic-wallaby' - Created task run 'async_task-0' for task 'async_task'
12:34:50.871 | INFO    | Flow run 'organic-wallaby' - Executing 'async_task-0' immediately...
12:34:50.895 | INFO    | Task run 'async_task-0' - <_MainThread(MainThread, started 8725699392)>
12:34:50.911 | INFO    | Task run 'async_task-0' - Finished in state Completed()
12:34:50.924 | INFO    | Flow run 'organic-wallaby' - Created task run 'sync_task-1' for task 'sync_task'
12:34:50.925 | INFO    | Flow run 'organic-wallaby' - Submitted task run 'sync_task-1' for execution.
12:34:50.935 | INFO    | Flow run 'organic-wallaby' - Created task run 'async_task-1' for task 'async_task'
12:34:50.935 | INFO    | Flow run 'organic-wallaby' - Submitted task run 'async_task-1' for execution.
12:34:50.958 | INFO    | Task run 'sync_task-1' - <Thread(WorkerThread-0, started 6246838272)>
12:34:50.973 | INFO    | Task run 'async_task-1' - <Thread(WorkerThread-1, started 10804703232)>
12:34:50.977 | INFO    | Task run 'sync_task-1' - Finished in state Completed()
12:34:50.988 | INFO    | Task run 'async_task-1' - Finished in state Completed()
12:34:51.003 | INFO    | Flow run 'organic-wallaby' - Finished in state Completed('All states completed.')
---
12:34:51.071 | INFO    | prefect.engine - Created flow run 'melodic-bullfinch' for flow 'async-flow'
12:34:51.115 | INFO    | Flow run 'melodic-bullfinch' - <_MainThread(MainThread, started 8725699392)>
12:34:51.130 | INFO    | Flow run 'melodic-bullfinch' - Created task run 'sync_task-0' for task 'sync_task'
12:34:51.130 | INFO    | Flow run 'melodic-bullfinch' - Executing 'sync_task-0' immediately...
12:34:51.154 | INFO    | Task run 'sync_task-0' - <Thread(WorkerThread-2, started 10771050496)>
12:34:51.174 | INFO    | Task run 'sync_task-0' - Finished in state Completed()
12:34:51.186 | INFO    | Flow run 'melodic-bullfinch' - Created task run 'sync_task-1' for task 'sync_task'
12:34:51.187 | INFO    | Flow run 'melodic-bullfinch' - Submitted task run 'sync_task-1' for execution.
12:34:51.193 | INFO    | Flow run 'melodic-bullfinch' - Created task run 'async_task-0' for task 'async_task'
12:34:51.193 | INFO    | Flow run 'melodic-bullfinch' - Executing 'async_task-0' immediately...
12:34:51.219 | INFO    | Task run 'sync_task-1' - <Thread(WorkerThread-3, started 6263664640)>
12:34:51.231 | INFO    | Task run 'async_task-0' - <_MainThread(MainThread, started 8725699392)>
12:34:51.239 | INFO    | Task run 'sync_task-1' - Finished in state Completed()
12:34:51.248 | INFO    | Task run 'async_task-0' - Finished in state Completed()
12:34:51.259 | INFO    | Flow run 'melodic-bullfinch' - Created task run 'async_task-1' for task 'async_task'
12:34:51.259 | INFO    | Flow run 'melodic-bullfinch' - Submitted task run 'async_task-1' for execution.
12:34:51.282 | INFO    | Task run 'async_task-1' - <_MainThread(MainThread, started 8725699392)>
12:34:51.295 | INFO    | Task run 'async_task-1' - Finished in state Completed()
12:34:51.309 | INFO    | Flow run 'melodic-bullfinch' - Finished in state Completed('All states completed.')
```

Now possible, previously would fail with event loop errors

```python
from prefect import flow, task
import asyncio
import httpx


@task()
async def query_api(client: httpx.AsyncClient):
    resp = await client.get("data")
    return resp


@flow()
async def my_flow():
    async with httpx.AsyncClient(base_url="https://example.com") as client:
        print(await query_api(client))


if __name__ == "__main__":
    asyncio.run(my_flow())
```

### CODE REVIEW

File: src/prefect/_internal/concurrency/api.py [hunk 1]
```diff
@@ -231,6 +234,25 @@ def wait_for_call_in_loop_thread(
...
+                        done_callbacks=done_callbacks,
+                        contexts=contexts,
+                    )
+                )
+            else:
+                logger.error(
+                    "Detected unsafe call to `from_sync` from thread with event loop. "
+                    "Use `await greenback.ensure_portal()` to allow call to run "
+                    "without blocking the event loop."
+                )
```
Comment (zanieb) [inline src/prefect/_internal/concurrency/api.py:254]: I think this should never happen in production, just during internal changes. Maybe we should throw an exception here?

**Comment by zanieb** [comment]
This fails with following test case

```python
import threading

from prefect import flow, task


@task
def sync_task():
    print(threading.current_thread())


@task
async def async_task():
    print(threading.current_thread())


@flow(log_prints=True)
def sync_flow():
    print(threading.current_thread())

    # Run the async flow within the synchronous flow
    async_flow()  # Runs on worker thread 0


@flow(log_prints=True)
async def async_flow():
    print(threading.current_thread())

    await (
        async_task.submit()
    )  # Runs on worker thread 1 on `main` but runs in worker thread 0 on branch


# Because the flow function exits without waiting for the task above to finish,
# an error is thrown since worker thread 0 does not exist anymore

sync_flow()
```

**Comment by zanieb** [comment]
Here's an example for the "subflow" case

```python
# This example illustrates execution threads of nested flows and tasks

import asyncio
import threading

from prefect import flow, task


@task
def sync_task():
    print(threading.current_thread())


@task
async def async_task():
    print(threading.current_thread())


@flow(log_prints=True)
def sync_flow():
    print(threading.current_thread())

    sync_task()  # Runs on main thread
    async_task()  # Runs on main thread
    sync_task.submit()  # Runs on worker thread
    async_task.submit()  # Runs on worker thread

    async_subflow()  # Runs on main thread


@flow(log_prints=True)
async def async_subflow():
    print(threading.current_thread())

    sync_task()  # Runs on worker thread (3)
    sync_task.submit()  # Runs on worker thread
    await async_task()  # Runs on main thread
    await async_task.submit()  # Runs on main thread


sync_flow()
print("---")


@flow(log_prints=True)
async def async_flow():
    print(threading.current_thread())

    sync_task()  # Runs on worker thread
    sync_task.submit()  # Runs on worker thread
    await async_task()  # Runs on main thread
    await async_task.submit()  # Runs on main thread

    sync_subflow()  # Runs on worker thread


@flow(log_prints=True)
def sync_subflow():
    print(threading.current_thread())

    sync_task()  # Runs on current worker thread
    sync_task.submit()  # Runs on new worker thread
    async_task()  # Runs on current worker thread
    async_task.submit()  # Runs on new worker thread


asyncio.run(async_flow())
```

**Comment by zanieb** [comment]
Thanks! Feel free to do what you want with it of course, I just got excited after seeing that it was unblocked upstream :) I'm happy to try to answer any questions.

**Comment by zzstoatzz** [comment]
> although I am little rusty these days.

*ba dum hiss*

**Comment by cpdeethree** [comment]
Thank you for carrying this to the end @zanieb. This is excellent and super happy to get to see this fundamental change that can unlock a lot more use cases in prefect come in

**Review by chrisguidry — APPROVED** [review]
Folks are stoked for this one, Zanie!  Thanks again!

**Comment by zanieb** [comment]
Whoa that was quick! Exciting to see it merged <3 I hope this unblocks lots of use-cases

**Pull request merged**